kubetest2/kubetest2-gke/deployer/deployer.go

```diff
@@ -0,0 +1,516 @@
+	fmt.Printf("Gcloud command: gcloud %+v", args)
+	if err := runWithOutput(exec.Command("gcloud", args...)); err != nil {
+		return fmt.Errorf("error creating cluster: %v", err)
+	}
+	return nil
+}
+
+func (d *deployer) IsUp() (up bool, err error) {
+	// naively assume that if the api server reports nodes, the cluster is up
+	lines, err := exec.CombinedOutputLines(
+		exec.Command("kubectl", "get", "nodes", "-o=name"),
+	)
+	if err != nil {
+		return false, metadata.NewJUnitError(err, strings.Join(lines, "\n"))
+	}
+	return len(lines) > 0, nil
+}
+
+// DumpClusterLogs for GKE generates a small script that wraps
+// log-dump.sh with the appropriate shell-fu to get the cluster
+// dumped.
+//
+// TODO(RonWeber): This whole path is really gross, but this seemed
+// the least gross hack to get this done.
+//
+// TODO(RonWeber): Make this work with multizonal and regional clusters.
+func (d *deployer) DumpClusterLogs() error {
+	// gkeLogDumpTemplate is a template of a shell script where
+	// - %[1]s is the project
+	// - %[2]s is the zone
+	// - %[3]s is a filter composed of the instance groups
+	// - %[4]s is the log-dump.sh command line
+	const gkeLogDumpTemplate = `
+function log_dump_custom_get_instances() {
+  if [[ $1 == "master" ]]; then
+    return 0
+  fi
+
+  gcloud compute instances list '--project=%[1]s' '--filter=%[4]s' '--format=get(name)'
+}
+export -f log_dump_custom_get_instances
+# Set below vars that log-dump.sh expects in order to use scp with gcloud.
+export PROJECT=%[1]s
+export ZONE='%[2]s'
+export KUBERNETES_PROVIDER=gke
+export KUBE_NODE_OS_DISTRIBUTION='%[3]s'
+%[5]s
+`
+	// Prevent an obvious injection.
+	if strings.Contains(d.localLogsDir, "'") || strings.Contains(d.gcsLogsDir, "'") {
+		return fmt.Errorf("%q or %q contain single quotes - nice try", d.localLogsDir, d.gcsLogsDir)
+	}
+
+	// Generate a slice of filters to be OR'd together below
+	if err := d.getInstanceGroups(); err != nil {
+		return err
+	}
+	var filters []string
+	for _, ig := range d.instanceGroups {
+		filters = append(filters, fmt.Sprintf("(metadata.created-by:*%s)", ig.path))
+	}
+
+	// Generate the log-dump.sh command-line
+	dumpCmd := fmt.Sprintf("./cluster/log-dump/log-dump.sh '%s'", d.localLogsDir)
+	if d.gcsLogsDir != "" {
+		dumpCmd += " " + d.gcsLogsDir
+	}
+
+	return runWithOutput(exec.Command("bash", "-c", fmt.Sprintf(gkeLogDumpTemplate,
+		d.project,
+		d.zone,
+		os.Getenv("NODE_OS_DISTRIBUTION"),
+		strings.Join(filters, " OR "),
+		dumpCmd)))
+}
+
+func (d *deployer) TestSetup() error {
+	if d.testPrepared {
+		// Ensure setup is a singleton.
+		return nil
+	}
+	if err := d.getKubeConfig(); err != nil {
+		return err
+	}
+	if err := d.getInstanceGroups(); err != nil {
+		return err
+	}
+	if err := d.ensureFirewall(); err != nil {
+		return err
+	}
+	d.testPrepared = true
+	return nil
+}
+
+func (d *deployer) getKubeConfig() error {
+	info, err := os.Stat(d.kubecfg)
+	if err != nil {
+		return err
+	}
+	if info.Size() > 0 {
```

Comment:
this check seems strange

Reply:
There's an almost-logical explanation: Somewhere far away, we use ioutil.TempFile to set d.kubecfg to a filename we can safely overwrite.  But ioutil.TempFile creates the file.  So we leave it empty and make our real existence check "size > 0".

Reply:
we should probably refactor that to a tempdir :-)